src/ast/nodes/ObjectExpression.ts

```diff
@@ -54,6 +54,16 @@ export default class ObjectExpression extends NodeBase implements DeoptimizableE
 		super.bind();
 		// ensure the propertyMap is set for the tree-shaking passes
 		this.getPropertyMap();
+		if (
+			!this.hasUnknownDeoptimizedProperty &&
+			this.properties.some(prop => {
+				if (prop instanceof SpreadElement || prop.computed) return false;
+				if (prop.key instanceof Identifier) return prop.key.name == "__proto__";
+				return String((prop.key as Literal).value) == "__proto__";
+			})
```

Comment:
This could actually be simplified by using the property map:

```suggestion
		const propertyMap = this.getPropertyMap();
		if (
			!this.hasUnknownDeoptimizedProperty &&
			propertyMap.__proto__?.exactMatchRead
```

that means that there is at least one property in the object that can be explicitly derived to be "__proto__" (and is not a setter). That would also automatically cover cases like this:

```js
var foo = {['__proto__']: obj}

const getProto () => '__proto__' ;
var foo = {[getProto()]: obj}
```

One thing that is not covered, though is an unknown computed property that resolves to __proto__. If we would want to include that as well, then it gets kind of complicated.

Another alternative here would be to integrate this into `getPropertyMap`. Because in the end, `__proto__` does not deoptimize everything, it just changes the prototype so that there may be additional "hidden" properties. So it would be equivalent to the very first property being an "unknown computed" property: All properties below are still valid.
Simplifying it further, one could just treat "__proto__" like an unknown computed property when building the property map. Thoughts?

Reply:
Computed properties don't work for this feature (they create a regular property named `__proto__`, don't set the prototype), so that's not an issue.

This could definitely be smarter, this patch mostly tries to crudely patch the soundness hole. I could switch it over to using an unknown computed property, if you think that works. I'll look into it next week.

Reply:
TIL: `__proto__` needs to be a written string in object expressions. Ouch. Thanks.

Reply:
I've pushed a patch that moves this to `getPropertyMap` as suggested.